### ISSUE TITLE
urql@3.0.4 does not work with GCG typescript-urql plugin

### ISSUE THREAD

**Issue by esistgut** [urql-graphql/urql#3026]
### Describe the bug

Repository showing the issue: https://github.com/esistgut/testurql
check generated.ts, there is the following error on line 1325:
`Argument of type '{ requestPolicy?: Urql.RequestPolicy | undefined; context?: Partial<Urql.OperationContext> | undefined; pause?: boolean | undefined; variables?: Exact<{ [key: string]: never; }> | undefined; query: DocumentNode; }' is not assignable to parameter of type 'UseQueryArgs<Exact<{ [key: string]: never; }>, AllFilmsQuery>'.
  Type '{ requestPolicy?: Urql.RequestPolicy | undefined; context?: Partial<Urql.OperationContext> | undefined; pause?: boolean | undefined; variables?: Exact<{ [key: string]: never; }> | undefined; query: DocumentNode; }' is not assignable to type '{ requestPolicy?: RequestPolicy | undefined; context?: Partial<OperationContext> | undefined; pause?: boolean | undefined; } & { query: string | ... 1 more ... | TypedDocumentNode<...>; variables: Exact<...>; }'.
    Type '{ requestPolicy?: Urql.RequestPolicy | undefined; context?: Partial<Urql.OperationContext> | undefined; pause?: boolean | undefined; variables?: Exact<{ [key: string]: never; }> | undefined; query: DocumentNode; }' is not assignable to type '{ query: string | DocumentNode | TypedDocumentNode<AllFilmsQuery, Exact<{ [key: string]: never; }>>; variables: Exact<{ [key: string]: never; }>; }'.
      Types of property 'variables' are incompatible.
        Type 'Exact<{ [key: string]: never; }> | undefined' is not assignable to type 'Exact<{ [key: string]: never; }>'.
          Type 'undefined' is not assignable to type 'Exact<{ [key: string]: never; }>'.ts(2345)`
if you change urql version to 3.0.3 and do an `npm i` the error disappears.

### Reproduction

https://github.com/esistgut/testurql

### Urql version

urql 3.0.4

### Validations

- [X] I can confirm that this is a bug report, and not a feature request, RFC, question, or discussion, for which [GitHub Discussions](https://github.com/urql-graphql/urql/discussions) should be used
- [X] Read the [docs](https://formidable.com/open-source/urql/docs/).
- [X] Follow our [Code of Conduct](https://github.com/urql-graphql/urql/blob/main/CODE_OF_CONDUCT.md)

**Comment by kitten** [comment]
If I'm understanding this correctly, this is using generators that generate wrapper hooks for React and specifically for the urql (i.e. react-ueql) package.

To keep things short, this is the issue: `variables?: Exact<{ [key: string]: never; }> | undefined;` and this is obviously not matching. I mean, there could be more issues.

Anyway, this codegen plugin isn't being maintained by the GraphQL Code Generator project any longer, and there's several reasons for that which we don't really have to get into right now 😅 It's also something we don't and can't support. While we have ambitions to be a little more active in providing a TypeScript plugin/generator of our own in the future — which would be officially supported — as you can see, we can't properly maintain type compatibility with these specific type patterns indefinitely, especially if we're not aware of them

So, tl;dr please switch to using one of their approaches that generate `TypedDocumentNode`. They have some great documentation on the effort, and while it's maybe not always advertised everywhere, `TypedDocumentNode` generation is superior in all ways to generating hooks or function wrappers.

It works with GraphQL files, inline strings, tagged template literals, etc; there's several ways of generating them with GCG.
Best of all, they just wrap around `DocumentNode`s and add generic types, so wherever you pass them to urql, `@urql/core`, or any other API we maintain, their types will infer other types in our API ✌️

Hope that makes sense. I'll still leave this open for a bit as a notice, but I'll rename the title for clarity.
Let me know if you have any questions ♥️

**Comment by esistgut** [comment]
Hi @kitten,
I'm not sure what you mean by `TypedDocumentNode` approach, I searched the documentation and found https://the-guild.dev/graphql/codegen/plugins/typescript/typed-document-node but this seems to be deprecated, as many other plugins, in favor of the `client-preset` plugin.
The `client-preset` approach is the one used in the urql documentation too so I reworked my sample following https://formidable.com/open-source/urql/docs/basics/typescript-integration/, you can check it on the `testnew` branch: https://github.com/esistgut/testurql/tree/testnew
It appears to be still showing some kind of regression regarding "variables" in 3.0.4:
```
Argument of type '{ query: TypedDocumentNode<AllFilmsQuery, Exact<{ [key: string]: never; }>>; }' is not assignable to parameter of type 'UseQueryArgs<Exact<{ [key: string]: never; }>, AllFilmsQuery>'.
  Type '{ query: TypedDocumentNode<AllFilmsQuery, Exact<{ [key: string]: never; }>>; }' is not assignable to type '{ requestPolicy?: RequestPolicy | undefined; context?: Partial<OperationContext> | undefined; pause?: boolean | undefined; } & { query: string | ... 1 more ... | TypedDocumentNode<...>; variables: Exact<...>; }'.
    Property 'variables' is missing in type '{ query: TypedDocumentNode<AllFilmsQuery, Exact<{ [key: string]: never; }>>; }' but required in type '{ query: string | DocumentNode | TypedDocumentNode<AllFilmsQuery, Exact<{ [key: string]: never; }>>; variables: Exact<{ [key: string]: never; }>; }'.ts(2345)
urql-core-chunk.d.ts(751, 5): 'variables' is declared here.
```
this disappears as soon as I switch to version 3.0.3.
I can work around passing an empty variables object (`variables: {}`) but this doesn't seem right.

**Comment by esistgut** [comment]
Please note that my last sample does not use `typescript-urql` any longer.

**Comment by kitten** [comment]
- `@graphql-codegen/typed-document-node` will still just work, for instance,
- `@graphql-codegen/client-preset` will also generate `TypedDocumentNode`s so it'll also work
So, both are supported because they interact with `urql` via the `TypedDocumentNode` type.

However, settings the `Variables` generic to `Exact<{ [key: string]: never; }>` is not supported. I have issued a patch for this anyway in #3029.
That said though, even with this patch, I cannot guarantee that `@graphql-codegen/typescript-urql` will work or will continue to work.

We've basically never really recommended or adopted support for it and code generators that generate wrappers. The types generally don't work out, and we can't treat internal type structures as public interfaces; instead, only exposed types (and not how they internally work) are public API surface for us. Hence, the need and support for `TypedDocumentNode`, which has now been around for a while.

So, tl;dr I'll treat #3029 as a resolution of this issue, but I can't 100% guarantee that this fixes your issue. Even if it doesn't however, we can't treat this as our own issue anymore, as we can't influence the types generated by another tool.

**Comment by kitten** [comment]
@esistgut: Yep, you're right; just tested it and `Exact<{ [key: string]: never; }>` is still generated by `client-preset` at least, so the fix should take care of that. I think though that `typescript-urql` will still fail, since it's just too reliant on our types working a certain way

**Comment by kitten** [comment]
Should be [resolved in `@urql/core@3.2.2`](https://github.com/urql-graphql/urql/releases/tag/%40urql%2Fcore%403.2.2). Just make sure that you deduplicate `@urql/core`, as per usual, so you don't duplicate versions and hence keep the old type around

**Comment by esistgut** [comment]
I can confirm `@urql/core@3.2.2` fixes both the `client-preset` and the old `typescript-urql` approaches.
